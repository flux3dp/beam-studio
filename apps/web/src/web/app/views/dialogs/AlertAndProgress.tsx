import React, { useContext, useEffect, useRef } from 'react';

import Alert from 'app/views/dialogs/Alert';
import browser from 'implementations/browser';
import NonStopProgress from 'app/views/dialogs/NonStopProgress';
import Progress from 'app/views/dialogs/Progress';
import ProgressConstants from 'app/constants/progress-constants';
import { AlertProgressContext } from 'app/contexts/AlertProgressContext';
import { IAlert } from 'interfaces/IAlert';
import { IProgressDialog } from 'interfaces/IProgress';

const isProgress = (d: IAlert | IProgressDialog): d is IProgressDialog => d.isProgress;

const AlertsAndProgress = (): JSX.Element => {
  const messageRef = useRef<HTMLPreElement>();

  const { alertProgressStack } = useContext(AlertProgressContext);

  useEffect(() => {
    const message = messageRef.current as Element;
    if (message) {
      const aElements = message.querySelectorAll('a');
      for (let i = 0; i < aElements.length; i += 1) {
        const a = aElements[i];
        a.addEventListener('click', (e) => {
          e.preventDefault();
          browser.open(a.getAttribute('href'));
        });
      }
    }
  });
  if (alertProgressStack.length === 0) return <div />;

  const alertModals = alertProgressStack.map((data) => {
    if (isProgress(data)) {
      if (data.type === ProgressConstants.NONSTOP) {
        return <NonStopProgress key={`${data.key}-${data.id}`} data={data} />
      }
      return <Progress key={`${data.key}-${data.id}`} data={data} />;
    }

    return <Alert key={`${data.key}-${data.id}`} data={data} />;
  });

  return <div>{alertModals}</div>;
};

export default AlertsAndProgress;
