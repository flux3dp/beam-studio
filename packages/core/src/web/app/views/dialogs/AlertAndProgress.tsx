import React, { useContext, useEffect, useRef } from 'react';

import ProgressConstants from '@core/app/constants/progress-constants';
import { AlertProgressContext } from '@core/app/contexts/AlertProgressContext';
import Alert from '@core/app/views/dialogs/Alert';
import NonStopProgress from '@core/app/views/dialogs/NonStopProgress';
import Progress from '@core/app/views/dialogs/Progress';
import type { IAlert } from '@core/interfaces/IAlert';
import type { IProgressDialog } from '@core/interfaces/IProgress';

import browser from '@app/implementations/browser';

const isProgress = (d: IAlert | IProgressDialog): d is IProgressDialog => d.isProgress;

const AlertsAndProgress = (): React.JSX.Element => {
  const messageRef = useRef<HTMLPreElement>();

  const { alertProgressStack } = useContext(AlertProgressContext);

  useEffect(() => {
    const message = messageRef.current as Element;

    if (message) {
      const aElements = message.querySelectorAll('a');

      for (let i = 0; i < aElements.length; i += 1) {
        const a = aElements[i];

        a.addEventListener('click', (e) => {
          e.preventDefault();
          browser.open(a.getAttribute('href'));
        });
      }
    }
  });

  if (alertProgressStack.length === 0) {
    return <div />;
  }

  const alertModals = alertProgressStack.map((data) => {
    if (isProgress(data)) {
      if (data.type === ProgressConstants.NONSTOP) {
        return <NonStopProgress data={data} key={`${data.key}-${data.id}`} />;
      }

      return <Progress data={data} key={`${data.key}-${data.id}`} />;
    }

    return <Alert data={data} key={`${data.key}-${data.id}`} />;
  });

  return <div>{alertModals}</div>;
};

export default AlertsAndProgress;
