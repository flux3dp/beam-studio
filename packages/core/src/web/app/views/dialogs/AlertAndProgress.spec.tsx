import React from 'react';

import { render } from '@testing-library/react';

import AlertAndProgress from './AlertAndProgress';

jest.mock('@core/app/views/dialogs/Alert', () => ({ data }: any) => <div>DummyAlert {JSON.stringify(data)}</div>);
jest.mock('@core/app/views/dialogs/Progress', () => ({ data }: any) => <div>DummyProgress {JSON.stringify(data)}</div>);
jest.mock('@core/app/views/dialogs/NonStopProgress', () => ({ data }: any) => (
  <div>DummyNonStopProgress {JSON.stringify(data)}</div>
));

jest.mock('@core/app/contexts/AlertProgressContext', () => ({
  AlertProgressContext: React.createContext({
    alertProgressStack: [
      {
        buttons: [
          {
            label: 'Yes',
            title: 'Yes',
          },
          {
            label: 'No',
            title: 'No',
          },
        ],
        caption: 'Hello World',
        checkboxCallbacks: jest.fn(),
        checkboxText: 'Select',
        iconUrl: 'https://www.flux3dp.com/icon.svg',
        id: 'alert',
        message: 'Yes or No',
      },
      {
        id: 'progress',
        isProgress: true,
      },
    ],
    popById: jest.fn(),
    popFromStack: jest.fn(),
  }),
}));

test('should render correctly', () => {
  const { baseElement } = render(<AlertAndProgress />);

  expect(baseElement).toMatchSnapshot();
});
