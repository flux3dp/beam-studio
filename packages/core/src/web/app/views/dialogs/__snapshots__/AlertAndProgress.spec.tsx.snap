// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<body>
  <div>
    <div>
      <div>
        DummyAlert 
        {"buttons":[{"label":"Yes","title":"Yes"},{"label":"No","title":"No"}],"caption":"Hello World","checkboxText":"Select","iconUrl":"https://www.flux3dp.com/icon.svg","id":"alert","message":"Yes or No"}
      </div>
      <div>
        DummyProgress 
        {"id":"progress","isProgress":true}
      </div>
    </div>
  </div>
</body>
`;
